content/2016-07-05-pythran-practice.md

```diff
@@ -0,0 +1,433 @@
+```
+
+This recursive version need a global variable to cache data as Python is limited by the recursion depth.
+As Pythran can't handle non constant globals, we will pass this variable as a second argument
+
+Now time it
+
+```bash
+python -c "import prob_14; print prob_14.run()"
+python -m timeit -s "import prob_14" "prob_14.run()"
+```
+
+```bash
+837799
+10 loops, best of 3: 1.26 msec per loop
+```
+<!-- It is 189 msec with the globals variable but I don't understand why it is so fast... -->
+
+OK, it will be the reference.
+
+## Imperative solution
+
+We can perform the direct computation
+
+```python
+>>> def collatz_chain_length(x):
+>>>     count = 0
+>>>     while x > 1:
+>>>         if x % 2 == 0:
+>>>             x = x / 2
+>>>         else:
+>>>             x = x * 3 + 1
+>>>         count += 1
+>>>     return count
+```
+
+Now time it
+
+```bash
+python -c "import prob_14; print prob_14.run()"
+python -m timeit -s "import prob_14" "prob_14.run()"
+```
+
+```bash
+837799
+10 loops, best of 3: 12.8 sec per loop
+```
+
+This solution is slower as it compute everything for each value (no caching mechanism)
+
+## Pythranize the solution
+
+Look at the imperative solution. We just add in the .py file:
+
+`# pythran export run()`
+
+```bash
+python -m pythran.run prob_14.py
+python -c "import prob_14; print prob_14.run()"
+python -m timeit -s "import prob_14" "prob_14.run()"
+```
+
+```
+837799
+10 loops, best of 3: 626 msec per loop
+```
+
+Result is the same (hopefully...) but it is 20x faster with no effort.
+As we use native library, we can go without the [GIL](https://wiki.python.org/moin/GlobalInterpreterLock)
+and use Pythran OpenMP extension:
+
+```python
+>>> # pythran export run()
+>>> def run():
+>>>     max_k, max_v = 1, 0
+>>>     # omp parallel for
+>>>     for i in xrange(1, 1000000):
+>>>         new_len = collatz_chain_length2(i)
+>>>         # omp critical
+>>>         if new_len > max_v:
+>>>             max_k, max_v = i, new_len
+>>>     return max_k
+```
+
+```bash
+python -m pythran.run prob_14.py -fopenmp
+python -c "import prob_14; print prob_14.run()"
+python -m timeit -s "import prob_14" "prob_14.run()"
+```
+
+```bash
+837799
+10 loops, best of 3: 210 msec per loop
+```
+
+My computer is have 4 CPU and there is another 3x speed up. Which is finally a nice 60x speed up.
+It is good but still slower than the recursive function as it doesn't cache values.
+
+
+Let's look at the recursive solution.
+
+```bash
+python -m pythran.run prob_14.py
+python -c "import prob_14; print prob_14.run()"
+python -m timeit -s "import prob_14" "prob_14.run()"
+```
+
+```bash
+837799
+10 loops, best of 3: 968 msec per loop
+```
+
+We get a 1.3x speed up which is almost nothing. This example show that Python is
+not so bad with high level construct like dictionary. As we spend most of the
+time looking and hashing data in dictionary, it is already a native code in Python
+and we are not really faster with Pythran.
```

Comment:
It will not help as we will need to sync between each task as datasharing only capture local variable so we need to execute the task before the end of the function.